torchvision/prototype/features/_dataset_wrapper.py

```diff
@@ -0,0 +1,388 @@
+        self.vision_dataset = dataset
+        self.sample_wrappers = wrappers
+
+        # We need to disable the transforms on the dataset here to be able to inject the wrapping before we apply the
+        # transforms
+        self.transform, dataset.transform = dataset.transform, None
+        self.target_transform, dataset.target_transform = dataset.target_transform, None
+        self.transforms, dataset.transforms = dataset.transforms, None
+
+    def __getattr__(self, item: str) -> Any:
+        with contextlib.suppress(AttributeError):
+            return object.__getattribute__(self, item)
+
+        return getattr(self.vision_dataset, item)
+
+    def __getitem__(self, idx: int) -> Any:
+        # This gets us the raw sample since we disabled the transforms for the underlying dataset in the constructor
+        # of this class
+        sample = self.vision_dataset[idx]
+
+        wrappers_flat, spec = tree_flatten(self.sample_wrappers)
+        # We cannot use `tree_flatten` directly, because the spec of `self.sample_wrappers` and `sample` might differ.
+        # For example, for `COCODetection` the target is a list of dicts. To be able to wrap this into a dict of lists,
+        # we need to have access to the whole list, but `tree_flatten` would also flatten it.
+        sample_flat = tree_flatten_to_spec(sample, spec)
+        wrapped_sample_flat = [wrapper(item) for item, wrapper in zip(sample_flat, wrappers_flat)]
+        sample = tree_unflatten(wrapped_sample_flat, spec)
+
+        # We don't need to care about `transform` and `target_transform` here since `VisionDataset` joins them into a
+        # `transforms` internally:
+        # https://github.com/pytorch/vision/blob/2d92728341bbd3dc1e0f1e86c6a436049bbb3403/torchvision/datasets/vision.py#L52-L54
+        if self.transforms is not None:
+            sample = self.transforms(*sample)
+
+        return sample
+
+    def __len__(self) -> int:
+        return len(self.vision_dataset)
+
+    @classmethod
+    def _register_wrappers_fn(cls, dataset_type: Type[datasets.VisionDataset]) -> Callable[[PyTree], PyTree]:
+        def register(wrappers_fn: PyTree) -> PyTree:
+            cls._wrappers_fns[dataset_type] = wrappers_fn
+            return wrappers_fn
+
+        return register
+
+    @classmethod
+    def from_torchvision_dataset(
+        cls,
+        dataset: datasets.VisionDataset,
+        *,
+        keep_pil_image: bool = False,
+        bounding_box_format: Optional[features.BoundingBoxFormat] = None,
+        dtypes: Optional[Dict[Type[features._Feature], Optional[torch.dtype]]] = None,
+    ) -> VisionDatasetFeatureWrapper:
+        dtypes: Dict[Type[features._Feature], Optional[torch.dtype]] = {
+            features.Image: torch.uint8,
+            features.Label: torch.int64,
+            features.Mask: torch.uint8,
+            features.BoundingBox: torch.float32,
+            features.GenericFeature: None,
+            **(dtypes or dict()),
+        }
```

Comment:
This gives users the ability to decide what dtype each feature should have. This is somewhat different than the `bounding_box_format` above, since the initial value is dictated by the dataset, whereas the dtype is often "free" since we are creating a tensor for the first time. By choosing the right dtype from the beginning, we can avoid copies later.

Reply:
> By choosing the right dtype from the beginning, we can avoid copies later.

For both Images and Masks which are the only datapoints that are somewhat expensive to convert, we still call `F.convert_image_dtype` under the hood so this isn't saving any copy.

In lack of clear performance benefit, I'd err towards letting users just use the `ToDtype()` transform?

Reply:
Yes. The only times this makes some sense is for datapoints that don't have a fixed starting dtype, e.g. bounding boxes which are often `List[int]` or `List[float]`. But, I agree, an extra copy of a `(N, 4)` tensor with `N < 20` is probably not something worth some extra API complexity.